ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/RootPeersDNS.hs

```diff
@@ -0,0 +1,461 @@
+    let taggedRelays = zipWith
+                        (\tag rel
+                              -> case rel of
+                                   RelayDomain (DomainAddress domain port)
+                                     -> RelayDomain (DomainAddress (domain <> (pack . show) tag) port)
+                                   x -> x
+                        )
+                        [(0 :: Int), 1 .. ]
+                        relays
+        relaysBlocks = blocks relaysPerGroup taggedRelays
+        relaysMap    = map Map.fromList $ zipWith zip relaysBlocks peerAdvertise
+
+        localRootPeers = zip targets relaysMap
+
+        domains = [ domain | RelayDomain (DomainAddress domain _) <- taggedRelays ]
+
+        ipsPerDomain = 2
+
+    ips <- blocks ipsPerDomain <$> vectorOf (ipsPerDomain * length domains) (toIPv4w <$> arbitrary)
+    gen <- mkStdGen <$> arbitrary
+
+    let dnsMap = Map.fromList $ zip domains ips
+
+    return (MockRoots {
+      mockLocalRootPeers = localRootPeers,
+      mockDNSMap         = dnsMap,
+      mockStdGen         = gen
+    })
+  where
+    blocks _ [] = []
+    blocks s l  = take s l : blocks s (drop s l)
+
+instance Arbitrary MockRoots where
+    arbitrary = genMockRoots
+
+-- | Used for debugging in GHCI
+--
+simpleMockRoots :: MockRoots
+simpleMockRoots = MockRoots localRootPeers dnsMap (mkStdGen 60)
+  where
+    localRootPeers = [ (2, Map.fromList [ (RelayAddress (read "192.0.2.1")           (read "3333") , DoAdvertisePeer)
+                                        , (RelayDomain  (DomainAddress "test.domain" (read "4444")), DoNotAdvertisePeer) ]
+                       )
+                  ]
+    dnsMap = Map.fromList [ ("test.domain", [read "192.1.1.1", read "192.2.2.2"])
+                          ]
+
+-- | Mock DNSActions data structure for testing purposes.
+-- Adds DNS Lookup function for IOSim with different timeout and lookup
+-- delays for every attempt.
+mockDNSActions :: forall exception s.
+               Map Domain [IPv4]
+               -> StrictTVar (IOSim s) StdGen
+               -> DNSActions () exception (IOSim s)
+mockDNSActions dnsMap stdGenVar = dnsActions
+ where
+   genDiffTime :: Integer -> Integer -> StdGen -> DiffTime
+   genDiffTime lo hi = picosecondsToDiffTime . fst . uniformR (lo * 1000000000, hi * 1000000000)
+
+
+   dnsResolverResource _ = return (constantResource ())
+   dnsAsyncResolverResource _ = return (constantResource ())
+
+   dnsLookupAWithTTL :: TimeoutFn (IOSim s)
+                     -> resolvConf
+                     -> resolver
+                     -> Domain
+                     -> IOSim s (Either DNSError [(IPv4, TTL)])
+   dnsLookupAWithTTL timeout _ _ domain = do
+     gen <- atomically $ readTVar stdGenVar
+
+         -- Small probability of timeout
+     let (dtTimeout, dtDelay) = (genDiffTime 110 300 gen, genDiffTime 20 120 gen)
+         (_, g') = split gen
+
+     atomically @(IOSim s) $ writeTVar stdGenVar g'
```

Comment:
^ it the type argument really needed here?

Reply:
Not needed! It was before some refactoring that I made. Thank you!